lib/decidim/spam_signal/spammer_signaler/spammer_signaler.rb

```diff
@@ -0,0 +1,114 @@
+require 'uri'
+
+module Decidim
+  module SpamSignal
+    class AppSpamSignal
+
+      def run!
+        @user_bot_email = ENV.fetch("USER_BOT_EMAIL", nil)
+        @uncheckable_email_domain = ENV.fetch("UNCHECKABLE_EMAIL_DOMAIN", nil)
+        @user_bot = take_or_create_user_bot
+        @regex = /#{take_regex}/i
+        
+        users_to_check.each do |user|
```

Comment:
How about using [`Decidim::UserModeration`](https://github.com/decidim/decidim/blob/4f8f8c47e340c95281712134652b298d28e31804/decidim-core/app/models/decidim/user_moderation.rb). 

Something like

```ruby
Decidim.UserModeration.unblocked.where(decidim_users: {email: <stuff>}) .includes(:reports).where("decidim_reports.moderation_id": nil).each do |moderation|
 user = moderation.user
 # do your checks
end
```

It won't really optimise the query (still in O(n)), but it's a bit better I think